diff --git a/docroot/modules/patched/entity_embed/src/Plugin/Filter/EntityEmbedFilter.php b/docroot/modules/patched/entity_embed/src/Plugin/Filter/EntityEmbedFilter.php
index 40de620..cb31276 100644
--- a/docroot/modules/patched/entity_embed/src/Plugin/Filter/EntityEmbedFilter.php
+++ b/docroot/modules/patched/entity_embed/src/Plugin/Filter/EntityEmbedFilter.php
@@ -126,8 +126,11 @@ public function process($text, $langcode) {
 
         $this->replaceNodeContent($node, $entity_output);
       }
-
-      $result->setProcessedText(Html::serialize($dom));
+      $processedText = Html::serialize($dom);
+      if (isset($context['data-button']) && $context['data-button']) {
+        $processedText = '<p class="button">' . Html::serialize($dom) . '</p>';
+      }
+      $result->setProcessedText($processedText);
     }
 
     return $result;
