diff --git a/includes/functions.php b/includes/functions.php
index 03f4709..a792ed8 100644
--- a/includes/functions.php
+++ b/includes/functions.php
@@ -93,8 +93,10 @@ if(!function_exists('strcode2utf')){
   function strcode2utf($str,$lo=true) {
 	//converts all the &#nnn; and &#xhhh; in a string to Unicode
 	if ($lo) { $lo = 1; } else { $lo = 0; }
-	$str = preg_replace('/\&\#([0-9]+)\;/me', "code2utf('\\1',{$lo})",$str);
-	$str = preg_replace('/\&\#x([0-9a-fA-F]+)\;/me', "codeHex2utf('\\1',{$lo})",$str);
+	$str = preg_replace_callback('/\&\#([0-9]+)\;/m', 
+		create_function ('$matches', 'return code2utf($matches[1], $lo);'), $str);
+	$str = preg_replace_callback('/\&\#x([0-9a-fA-F]+)\;/m', 
+		create_function ('$matches', 'return codeHex2utf($matches[1], $lo);'), $str);
 	return $str;
   }
 }
