diff --git a/datalayer.module b/datalayer.module
index e4eb570..badb9c8 100644
--- a/datalayer.module
+++ b/datalayer.module
@@ -325,7 +325,15 @@ function _datalayer_get_entity_data($entity) {
     // For entities with an owner/author, get the username.
     if ((empty($selected_properties) || array_key_exists('name', $properties))
       && !isset($output_data['entityName']) && is_object($entity->uid)) {
-      $output_data['entityName'] = $entity->uid->entity->name->getString();
+      if ($entity->uid->entity) {
+        if ($entity->uid->entity instanceof EntityInterface) {
+          $eentity = $entity->uid->entity;
+          $value = ['id' => $eentity->id(), 'label' => $eentity->label(), 'bundle' => $eentity->bundle()];
+          // Allow modules to alter entity reference data with hook_datalayer_entity_reference_alter().
+          \Drupal::moduleHandler()->alter('datalayer_entity_reference', $value, $eentity);
+          $output_data['entityName'] = $eentity->name->getString();
+        }
+      }
     }
 
     // Output group name.
@@ -558,7 +566,14 @@ function _datalayer_field_get_value($field_item, $field_type) {
   $value = [];
   switch ($field_type) {
     case 'entity_reference':
-      $value = $field_item->getValue() + ['target_name' => $field_item->entity->name->getString()];
+      if (!$field_item->isEmpty()) {
+        if ($field_item->entity instanceof EntityInterface) {
+          $entity = $field_item->entity;
+          $value = ['id' => $entity->id(), 'label' => $entity->label(), 'bundle' => $entity->bundle()];
+          // Allow modules to alter entity reference data with hook_datalayer_entity_reference_alter().
+          \Drupal::moduleHandler()->alter('datalayer_entity_reference', $value, $entity);
+        }
+      }
       break;
