diff --git a/docroot/modules/contrib/content_browser/content_browser.module b/docroot/modules/contrib/content_browser/content_browser.module
index 6da7c90..edbe740 100644
--- a/docroot/modules/contrib/content_browser/content_browser.module
+++ b/docroot/modules/contrib/content_browser/content_browser.module
@@ -109,10 +109,18 @@ function content_browser_form_views_exposed_form_alter(&$form, \Drupal\Core\Form
             $bundles = $settings['handler_settings']['target_bundles'];
 
             foreach ($form['type']['#options'] as $bundle => $label) {
-              if (!in_array($bundle, $bundles) && $bundle !== 'All') {
+              if (!in_array($bundle, $bundles)) {
                 unset($form['type']['#options'][$bundle]);
               }
             }
+
+            // Ensure that a proper default #value is set, to prevent form
+            // errors.
+            $input = $form_state->getUserInput();
+            if ($input['type'] == 'All') {
+              $input['type'] = reset($bundles);
+              $form_state->setUserInput($input);
+            }
           }
         }
       }
