diff --git a/docroot/modules/contrib/content_browser/content_browser.module b/docroot/modules/contrib/content_browser/content_browser.module
index 2aa8bb7..6da7c90 100644
--- a/docroot/modules/contrib/content_browser/content_browser.module
+++ b/docroot/modules/contrib/content_browser/content_browser.module
@@ -109,18 +109,10 @@ function content_browser_form_views_exposed_form_alter(&$form, \Drupal\Core\Form
             $bundles = $settings['handler_settings']['target_bundles'];
 
             foreach ($form['type']['#options'] as $bundle => $label) {
-              if (!in_array($bundle, $bundles)) {
+              if (!in_array($bundle, $bundles) && $bundle !== 'All') {
                 unset($form['type']['#options'][$bundle]);
               }
             }
-
-            // Ensure that a proper default #value is set, to prevent form
-            // errors.
-            if (!isset($form['type']['#value']) || $form['type']['#value'] == 'All') {
-              $form['type']['#value'] = reset($bundles);
-            }
-
-            $form['type']['#default_value'] = reset($bundles);
           }
         }
       }
