diff --git a/core/modules/node/src/NodeForm.php b/core/modules/node/src/NodeForm.php
index bb6c49d..88e7499 100644
--- a/core/modules/node/src/NodeForm.php
+++ b/core/modules/node/src/NodeForm.php
@@ -346,6 +346,8 @@ public function preview(array $form, FormStateInterface $form_state) {
     $store = $this->tempStoreFactory->get('node_preview');
     $this->entity->in_preview = TRUE;
     $store->set($this->entity->uuid(), $form_state);
+    // Avoid destination redirect.
+    \Drupal::request()->query->set('destination', '');
     $form_state->setRedirect('entity.node.preview', array(
       'node_preview' => $this->entity->uuid(),
       'view_mode_id' => 'default',
