diff --git a/js/geysir.js b/js/geysir.js
index f65c9ce..95c0bad 100644
--- a/js/geysir.js
+++ b/js/geysir.js
@@ -61,4 +61,11 @@
     Drupal.attachBehaviors();
   };

+  /**
+   * Triggered by AJAX action for page reload.
+   */
+  $.fn.reloadPageAjaxAction = function() {
+    location.reload();
+  };
+
 })(jQuery, Drupal, drupalSettings);
diff --git a/src/Form/GeysirModalParagraphForm.php b/src/Form/GeysirModalParagraphForm.php
index 5ec29ce..4ee19c2 100644
--- a/src/Form/GeysirModalParagraphForm.php
+++ b/src/Form/GeysirModalParagraphForm.php
@@ -104,6 +104,8 @@ class GeysirModalParagraphForm extends GeysirParagraphForm {
       $response->addCommand(new GeysirCloseModalDialogCommand());

       $response->addCommand(new GeysirReattachBehaviors());
+
+      $response->addCommand(new InvokeCommand(NULL, 'reloadPageAjaxAction'));
     }

     return $response;
